velox/functions/sparksql/Hash.cpp

```diff
@@ -265,6 +265,59 @@ class RowVectorHasher : public SparkVectorHasher<HashClass> {
   std::vector<std::shared_ptr<SparkVectorHasher<HashClass>>> hashers_;
 };
 
+template <typename HashClass, typename ReturnType, typename ArgType>
+void hashSimdTyped(
+    const SelectivityVector* rows,
+    std::vector<VectorPtr>& args,
+    FlatVector<ReturnType>& result) {
+  vector_size_t begin = rows->begin();
+  vector_size_t end = rows->end();
+  const __restrict ArgType* rawA =
+      args[0]->asUnchecked<FlatVector<ArgType>>()->rawValues();
+  auto* __restrict rawResult = result.template mutableRawValues<ReturnType>();
+  auto* __restrict rowsData = rows->allBits();
+  if (rows->isAllSelected()) {
```

Comment:
Just use `rows.applyToSelected`, that one is more efficient as it will check if each 64 bits word is dense so more simd opportunity.

Reply:
Thanks I replaced the logic in else branch with `applyToSelected`. I tested both logic 1 and 2 for the case that rows are all selected. The first one gives better results specific for short types.

1
```
  const ArgType* __restrict rawA =
      args[hashIdx]->asUnchecked<FlatVector<ArgType>>()->rawValues();
  auto* __restrict rawResult = result.template mutableRawValues<ReturnType>();
  auto begin = rows->begin();
  auto end = rows->end();
  if (rows->isAllSelected()) {
    for (auto row = begin; row < end; ++row) {
      rawResult[row] = hashOne<HashClass>(rawA[row], rawResult[row]);
    }
  } else {
    rows->applyToSelected([&](auto row) {
      rawResult[row] = hashOne<HashClass>(rawA[row], rawResult[row]);
    });
  }
```
Result
```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
hash#INTEGER#0%nulls##hash                                229.99us     4.35K
hash#INTEGER#0%nulls##xxhash64                            434.80us     2.30K
----------------------------------------------------------------------------
hash#REAL#0%nulls##hash                                   247.65us     4.04K
hash#REAL#0%nulls##xxhash64                               459.96us     2.17K
----------------------------------------------------------------------------
hash#SMALLINT#0%nulls##hash                               236.78us     4.22K
hash#SMALLINT#0%nulls##xxhash64                           427.68us     2.34K
----------------------------------------------------------------------------
hash#TINYINT#0%nulls##hash                                238.79us     4.19K
hash#TINYINT#0%nulls##xxhash64                            418.10us     2.39K
----------------------------------------------------------------------------
```

2
```
  const ArgType* __restrict rawA =
      args[hashIdx]->asUnchecked<FlatVector<ArgType>>()->rawValues();
  auto* __restrict rawResult = result.template mutableRawValues<ReturnType>();
  rows->applyToSelected([&](auto row) {
    rawResult[row] = hashOne<HashClass>(rawA[row], rawResult[row]);
  });
```
Result
```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
hash#INTEGER#0%nulls##hash                                444.32us     2.25K
hash#INTEGER#0%nulls##xxhash64                            432.89us     2.31K
----------------------------------------------------------------------------
hash#REAL#0%nulls##hash                                   541.78us     1.85K
hash#REAL#0%nulls##xxhash64                               461.43us     2.17K
----------------------------------------------------------------------------
hash#SMALLINT#0%nulls##hash                               441.32us     2.27K
hash#SMALLINT#0%nulls##xxhash64                           428.09us     2.34K
----------------------------------------------------------------------------
hash#TINYINT#0%nulls##hash                                441.84us     2.26K
hash#TINYINT#0%nulls##xxhash64                            418.52us     2.39K
----------------------------------------------------------------------------
```

Reply:
Can you try this?

```diff
 template <typename Callable>
 inline void SelectivityVector::applyToSelected(Callable func) const {
   if (isAllSelected()) {
-    for (vector_size_t row = begin_; row < end_; ++row) {
+    const auto end = end_;
+    for (vector_size_t row = begin_; row < end; ++row) {
       func(row);
     }
   } else {
```

Reply:
Yes, this makes difference. Thank you.